### PR TITLE
fix:(taps): querying only selected columns

### PR DESCRIPTION
**Pull request by drbjim**
making query statement with only selected columns.

### CODE REVIEW

**Comment by drbjim** [comment]
previous code pulls all columns from table even it's not selected from meltano-cli.
for that reason, I've made change to pull only selected columns to reduce resource for select command.

**Comment by visch** [comment]
@drbjim  I'm not seeing the behavior you are. 

Here's my table
image


Here's the data in the table
image


Here's my meltano.yml
```
version: 1
send_anonymous_usage_stats: true
project_id: "tap-postgres"
plugins:
  extractors:
  - name: "tap-postgres"
    namespace: "tap_postgres"
    pip_url: -e .
    capabilities:
    - state
    - catalog
    - discover
    config:
      sqlalchemy_url: "postgresql://postgres:postgres@localhost:5432/postgres"
    settings:
    - name: sqlalchemy_url
      kind: password
    - name: ssh_tunnel.enable
      kind: boolean
    - name: ssh_tunnel.private_key
      kind: password
    - name: ssh_tunnel.private_key_password
      kind: password
    - name: ssh_tunnel.host
    - name: ssh_tunnel.username
    - name: ssh_tunnel.port
    select:
    - "public-test.a"
```

Here's my running the job and it only outputing one column

```
visch@DESKTOP-9BDPA9T:~/git/tap-postgres$ meltano invoke tap-postgres
2023-05-04 14:44:17,791 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-_pg_foreign_data_wrappers'.
2023-05-04 14:44:17,791 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-_pg_foreign_servers'.
2023-05-04 14:44:17,791 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-_pg_foreign_table_columns'.
2023-05-04 14:44:17,792 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-_pg_foreign_tables'.
2023-05-04 14:44:17,792 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-_pg_user_mappings'.
2023-05-04 14:44:17,792 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-administrable_role_authorizations'.
2023-05-04 14:44:17,792 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-applicable_roles'.
2023-05-04 14:44:17,792 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-attributes'.
2023-05-04 14:44:17,792 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-character_sets'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-check_constraint_routine_usage'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-check_constraints'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-collation_character_set_applicability'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-collations'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-column_column_usage'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-column_domain_usage'.
2023-05-04 14:44:17,793 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-column_options'.
2023-05-04 14:44:17,794 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-column_privileges'.
2023-05-04 14:44:17,794 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-column_udt_usage'.
2023-05-04 14:44:17,794 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-columns'.
2023-05-04 14:44:17,794 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-constraint_column_usage'.
2023-05-04 14:44:17,795 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-constraint_table_usage'.
2023-05-04 14:44:17,795 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-data_type_privileges'.
2023-05-04 14:44:17,795 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-domain_constraints'.
2023-05-04 14:44:17,795 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-domain_udt_usage'.
2023-05-04 14:44:17,795 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-domains'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-element_types'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-enabled_roles'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-foreign_data_wrapper_options'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-foreign_data_wrappers'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-foreign_server_options'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-foreign_servers'.
2023-05-04 14:44:17,796 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-foreign_table_options'.
2023-05-04 14:44:17,797 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-foreign_tables'.
2023-05-04 14:44:17,797 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-information_schema_catalog_name'.
2023-05-04 14:44:17,797 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-key_column_usage'.
2023-05-04 14:44:17,797 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-parameters'.
2023-05-04 14:44:17,797 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-referential_constraints'.
2023-05-04 14:44:17,798 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-role_column_grants'.
2023-05-04 14:44:17,798 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-role_routine_grants'.
2023-05-04 14:44:17,798 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-role_table_grants'.
2023-05-04 14:44:17,798 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-role_udt_grants'.
2023-05-04 14:44:17,798 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-role_usage_grants'.
2023-05-04 14:44:17,798 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-routine_column_usage'.
2023-05-04 14:44:17,799 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-routine_privileges'.
2023-05-04 14:44:17,799 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-routine_routine_usage'.
2023-05-04 14:44:17,799 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-routine_sequence_usage'.
2023-05-04 14:44:17,799 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-routine_table_usage'.
2023-05-04 14:44:17,800 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-routines'.
2023-05-04 14:44:17,800 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-schemata'.
2023-05-04 14:44:17,800 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-sequences'.
2023-05-04 14:44:17,800 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-sql_features'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-sql_implementation_info'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-sql_parts'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-sql_sizing'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-table_constraints'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-table_privileges'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-tables'.
2023-05-04 14:44:17,801 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-transforms'.
2023-05-04 14:44:17,802 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-triggered_update_columns'.
2023-05-04 14:44:17,802 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-triggers'.
2023-05-04 14:44:17,802 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-udt_privileges'.
2023-05-04 14:44:17,802 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-usage_privileges'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-user_defined_types'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-user_mapping_options'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-user_mappings'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-view_column_usage'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-view_routine_usage'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-view_table_usage'.
2023-05-04 14:44:17,803 | INFO     | tap-postgres         | Skipping deselected stream 'information_schema-views'.
2023-05-04 14:44:17,804 | INFO     | tap-postgres         | Beginning full_table sync of 'public-test'...
2023-05-04 14:44:17,804 | INFO     | tap-postgres         | Tap has custom mapper. Using 1 provided map(s).
{"type": "SCHEMA", "stream": "public-test", "schema": {"properties": {"a": {"type": ["string", "null"]}}, "type": "object"}, "key_properties": []}
{"type": "RECORD", "stream": "public-test", "record": {"a": "a"}, "time_extracted": "2023-05-04T18:44:17.839174+00:00"}
```

**Comment by visch** [comment]
@drbjim Ok it is an efficiency gain you're going after I get it now! 

I'd like to get a test in place for this, something that confirms the computed SQL statement is what we'd expect with sqlalchemy when selecting a subset of the columns available on a table. I"m not sure off hand how easy sqlalchemy makes writing a test for this

Thank you very much @drbjim for putting this in :D

**Review by visch — CHANGES_REQUESTED** [review]
Let's get some tests in to verify the SQL is set to what we want. Mainly because I don't want future changes to break this behavior. Also the tests make it look like we need to run black on this code once or so (https://github.com/MeltanoLabs/tap-postgres/issues/98 would do this for us as well 🤷  )

**Comment by drbjim** [comment]
@visch Please run workflow again

File: tests/test_selected_schema.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+SAMPLE_CONFIG = {
+    "sqlalchemy_url": "postgresql://postgres:postgres@localhost:5432/postgres",
+}
+
+
+def selected_schema_test(tap, table_name):
+    """excluding one column from stream and check if it is not present in query"""
+    column_to_exclude = "name"
+    tap.run_discovery()
+    tap_catalog = json.loads(tap.catalog_json_text)
```
Comment (visch) [inline tests/test_selected_schema.py:19]: Seems like we're relying on another test to populate the table? We should just populate the DB with the table we care about and then just pull that table here so the test is isolated.

**Review by visch — CHANGES_REQUESTED** [review]
Looks really really good, just a few tweaks and I think we're there. Sorry this review took so long, trying to get all these reviews to happen at a minimum of weekly, haven't 100% committed yet but I"m close!

File: tests/test_selected_schema.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                metadata["metadata"]["selected"] = True
+                if metadata["breadcrumb"] != []:
+                    if metadata["breadcrumb"][1] == column_to_exclude:
+                        metadata["metadata"]["selected"] = False
+
+    tap = TapPostgres(config=SAMPLE_CONFIG, catalog=tap_catalog)
+    streams = tap.discover_streams()
+    selected_stream = [s for s in streams if s.selected is True][0]
+
+    assert not column_to_exclude in str(selected_stream.get_query())
```
Comment (visch) [inline tests/test_selected_schema.py:35]: Another approach here would be to look at the output from `tap-postgres` 🤷

**Comment by visch** [comment]
Great stuff @drbjim

**Review by visch — APPROVED** [review]

**Pull request merged**